indra/newview/llviewermessage.cpp

```diff
@@ -3240,159 +3224,140 @@ void send_agent_update(bool force_send, bool send_reliable)
+                send_update = true;
+                break;
             }
-            update_sec = cur_sec;
-            update_count = 0;
-        }
-        */
 
-        // Build the message
-        msg->newMessageFast(_PREHASH_AgentUpdate);
-        msg->nextBlockFast(_PREHASH_AgentData);
-        msg->addUUIDFast(_PREHASH_AgentID, gAgent.getID());
-        msg->addUUIDFast(_PREHASH_SessionID, gAgent.getSessionID());
-        msg->addQuatFast(_PREHASH_BodyRotation, body_rotation);
-        msg->addQuatFast(_PREHASH_HeadRotation, head_rotation);
-        msg->addU8Fast(_PREHASH_State, render_state);
-        msg->addU8Fast(_PREHASH_Flags, flags);
-
-//      if (camera_pos_agent.mV[VY] > 255.f)
-//      {
-//          LL_INFOS("Messaging") << "Sending camera center " << camera_pos_agent << LL_ENDL;
-//      }
-
-        msg->addVector3Fast(_PREHASH_CameraCenter, camera_pos_agent);
-        msg->addVector3Fast(_PREHASH_CameraAtAxis, LLViewerCamera::getInstance()->getAtAxis());
-        msg->addVector3Fast(_PREHASH_CameraLeftAxis, LLViewerCamera::getInstance()->getLeftAxis());
-        msg->addVector3Fast(_PREHASH_CameraUpAxis, LLViewerCamera::getInstance()->getUpAxis());
-        msg->addF32Fast(_PREHASH_Far, gAgentCamera.mDrawDistance);
+            // check camera rotation
+            // Note: we are using the sine small angle approximation trick here
+            constexpr F32 RADIANS_PER_DEGREE = F_PI / 360.f;
+            constexpr F32 CAMERA_AT_THRESHOLD = 0.2f * RADIANS_PER_DEGREE;
+            if ((last_camera_at - camera_at).magVec() > CAMERA_AT_THRESHOLD)
+            {
+                send_update = true;
+                break;
+            }
 
-        msg->addU32Fast(_PREHASH_ControlFlags, control_flags);
+            // check head rotation
+            constexpr F64 MIN_HEAD_ROT_QDOT = 0.9997;    // ~= 2.5 degrees -- if its less than this we need to update head_rot
+            constexpr F64 MAX_HEAD_ROT_QDOT = 0.99999;   // ~= 0.5 degrees -- if its greater than this then we consider it close enough
 
-        if (gDebugClicks)
-        {
-            if (control_flags & AGENT_CONTROL_LBUTTON_DOWN)
+            if (fabs((F64)(dot(last_body_rot, body_rotation))) < MIN_HEAD_ROT_QDOT)
             {
-                LL_INFOS("Messaging") << "AgentUpdate left button down" << LL_ENDL;
+                send_update = true;
+                break;
             }
 
-            if (control_flags & AGENT_CONTROL_LBUTTON_UP)
+            F64 head_rot_qdot = fabs((F64)(dot(last_head_rot, head_rotation)));
+            if (head_rot_qdot > MAX_HEAD_ROT_QDOT)
             {
-                LL_INFOS("Messaging") << "AgentUpdate left button up" << LL_ENDL;
+                // close enough
+                return;
+            }
+            else if (head_rot_qdot < MIN_HEAD_ROT_QDOT)
+            {
+                // way off
+                send_update = true;
+                break;
             }
```

Comment:
am I reading this wrong?  these conditions appear backwards

Reply:
For two quaternions that are very close to each other their dot-product will be very close to 1.0.  So:

(1) When the `dot_product` is above the `MAX `threshold then they are close --> no need to send an update at all.

(2) When the `dot_product `is less than the `MIN `threshold then definitely need to send an update.

(3) When the `dot_product` is between the two then maybe we need to send an update (to help mouselook aim) so we compute a "dynamic" threshold that moved from `MIN `to `MAX `over `THRESHOLD_GROWTH_PERIOD`.

I added the `fabs()` operation on the dot product because the quaternion hypersphere has a discrepancy of 2 when mapped to rotation space, so that handles the rare case when the two quaternions happen to be on opposite hemispheres.

Reply:
aha, ok thanks for the explanation